### ISSUE TITLE
Assertion `(IV)elen >= 0' failed.

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14481]
Migrated from [rt.perl.org#123755](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123755) (status was 'resolved')

Searchable as RT123755$

**Comment by p5pRT** [comment]
### From @hvds

#### Created by @hvds



Another issue found by AFL&#8203;:

% \./miniperl \-ce '/\(??/'
miniperl&#8203;: sv\.c&#8203;:11434&#8203;: Perl\_sv\_vcatpvfn\_flags&#8203;: Assertion \`\(IV\)elen >= 0' failed\.
Aborted \(core dumped\)
% 

I think the diff below is probably the right fix\, I'll apply it along with
a test in a couple of days if nobody disagrees\.

<details><summary>Inline Patch</summary>

```diff
diff --git a/regcomp.c b/regcomp.c
index 0d6d344..b5301ff 100644
--- a/regcomp.c
+++ b/regcomp.c
@@ -10176,7 +10176,8 @@ S_reg(pTHX_ RExC_state_t *pRExC_state, I32 paren, I32 *flagp,U32 depth)
 	    case '?':           /* (??...) */
 		is_logical = 1;
 		if (*RExC_parse != '{') {
-		    RExC_parse++;
+                    if (*RExC_parse)
+                        RExC_parse++;
                     /* diag_listed_as: Sequence (?%s...) not recognized in regex; marked by <-- HERE in m/%s/ */
                     vFAIL2utf8f(
                         "Sequence (%"UTF8f"...) not recognized",
```
</details>


<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.21.9:

Configured by hv at Fri Feb  6 18:25:03 GMT 2015.

Summary of my perl5 (revision 5 version 21 subversion 9) configuration:
  Commit id: fb0c7c3c47fb17bb0b685b957af9a0cc581c8996
  Platform:
    osname=linux, osvers=3.13.0-37-generic, archname=x86_64-linux
    uname='linux shad2 3.13.0-37-generic #64-ubuntu smp mon sep 22 21:28:38 utc 2014 x86_64 x86_64 x86_64 gnulinux '
    config_args='-des -Dcc=gcc -Dprefix=/opt/blead-d -Doptimize=-g -O6 -DDEBUGGING -Dusedevel -Uversiononly'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-fwrapv -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-g -O6',
    cppflags='-fwrapv -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.8.2', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678, doublekind=3
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16, longdblkind=3
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/gcc/x86_64-linux-gnu/4.8/include-fixed /usr/include/x86_64-linux-gnu /usr/lib /lib/x86_64-linux-gnu /lib/../lib /usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib
    libs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.19.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.19'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -g -O6 -L/usr/local/lib -fstack-protector'



@INC for perl 5.21.9:
    /opt/blead-d/lib/perl5/site_perl/5.21.9/x86_64-linux
    /opt/blead-d/lib/perl5/site_perl/5.21.9
    /opt/blead-d/lib/perl5/5.21.9/x86_64-linux
    /opt/blead-d/lib/perl5/5.21.9
    .


Environment for perl 5.21.9:
    HOME=/home/hv
    LANG=C
    LANGUAGE=en_GB:en
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/hv/bin:/home/hv/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @geeknik

Built v5\.21\.9 \(v5\.21\.8\-200\-ga57d3d4\) using the following command line&#8203;:

\./Configure \-des \-Dusedevel \-DDEBUGGING \-Dcc=afl\-gcc \-Doptimize=\-O2\\ \-g && AFL\_HARDEN=1 make \-j6 test\-prep

Bug found with AFL \(http&#8203;://lcamtuf\.coredump\.cx/afl\)

GDB output&#8203;:
gdb\-peda$ file ~/perl/perl
gdb\-peda$ set args test95\.pl
gdb\-peda$ r
\[Thread debugging using libthread\_db enabled\]
Using host libthread\_db library "/lib/x86\_64\-linux\-gnu/libthread\_db\.so\.1"\.
perl&#8203;: sv\.c&#8203;:11434&#8203;: Perl\_sv\_vcatpvfn\_flags&#8203;: Assertion \`\(IV\)elen >= 0' failed\.

Program received signal SIGABRT\, Aborted\.
\[\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-registers\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\]
RAX&#8203;: 0x0 
RBX&#8203;: 0x7fffffffe63a \-\-> 0x736574006c726570 \('perl'\)
RCX&#8203;: 0xffffffffffffffff 
RDX&#8203;: 0x6 
RSI&#8203;: 0x9c7c 
RDI&#8203;: 0x9c7c 
RBP&#8203;: 0x7ffff6ea9a07 \-\-> 0x257325732500203a \('&#8203;: '\)
RSP&#8203;: 0x7fffffffcde8 \-\-> 0x7ffff6d923e0 \(\<\*\_\_GI\_abort\+384>&#8203;:    mov    rdx\,QWORD PTR fs&#8203;:0x10\)
RIP&#8203;: 0x7ffff6d8f165 \(\<\*\_\_GI\_raise\+53>&#8203;:    cmp    rax\,0xfffffffffffff000\)
R8 : 0x7ffff7fdd700 \(0x00007ffff7fdd700\)
R9 : 0x56492860206e6f69 \('ion \`\(IV'\)
R10&#8203;: 0x8 
R11&#8203;: 0x202 
R12&#8203;: 0xf73ad8 \("\(IV\)elen >= 0"\)
R13&#8203;: 0xf899e0 \("Perl\_sv\_vcatpvfn\_flags"\)
R14&#8203;: 0x7ffff6ea9a07 \-\-> 0x257325732500203a \('&#8203;: '\)
R15&#8203;: 0x2caa
EFLAGS&#8203;: 0x202 \(carry parity adjust zero sign trap INTERRUPT direction overflow\)
\[\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-code\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\]
&nbsp;   0x7ffff6d8f15b \<\*\_\_GI\_raise\+43>&#8203;:    movsxd rdi\,eax
&nbsp;   0x7ffff6d8f15e \<\*\_\_GI\_raise\+46>&#8203;:    mov    eax\,0xea
&nbsp;   0x7ffff6d8f163 \<\*\_\_GI\_raise\+51>&#8203;:    syscall 
=> 0x7ffff6d8f165 \<\*\_\_GI\_raise\+53>&#8203;:    cmp    rax\,0xfffffffffffff000
&nbsp;   0x7ffff6d8f16b \<\*\_\_GI\_raise\+59>&#8203;:    ja     0x7ffff6d8f182 \<\*\_\_GI\_raise\+82>
&nbsp;   0x7ffff6d8f16d \<\*\_\_GI\_raise\+61>&#8203;:    repz ret 
&nbsp;   0x7ffff6d8f16f \<\*\_\_GI\_raise\+63>&#8203;:    nop
&nbsp;   0x7ffff6d8f170 \<\*\_\_GI\_raise\+64>&#8203;:    test   eax\,eax
\[\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-stack\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\]
0000| 0x7fffffffcde8 \-\-> 0x7ffff6d923e0 \(\<\*\_\_GI\_abort\+384>&#8203;:    mov    rdx\,QWORD PTR fs&#8203;:0x10\)
0008| 0x7fffffffcdf0 \-\-> 0xf73ad8 \("\(IV\)elen >= 0"\)
0016| 0x7fffffffcdf8 \-\-> 0x7ffff6eab9c1 \-\-> 0x706c6568007325 \('%s'\)
0024| 0x7fffffffce00 \-\-> 0x7fffffffce20 \-\-> 0x3000000018 
0032| 0x7fffffffce08 \-\-> 0x2caa 
0040| 0x7fffffffce10 \-\-> 0x7fffffffcf10 \-\-> 0x7fffffffe63a \-\-> 0x736574006c726570 \('perl'\)
0048| 0x7fffffffce18 \-\-> 0x7ffff6dc3fe6 \(\<\_\_fxprintf\+310>&#8203;:    lea    rsp\,\[rbp\-0x20\]\)
0056| 0x7fffffffce20 \-\-> 0x3000000018 
\[\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\]
Legend&#8203;: code\, data\, rodata\, value
Stopped reason&#8203;: SIGABRT
0x00007ffff6d8f165 in \*\_\_GI\_raise \(sig=\<optimized out>\) at \.\./nptl/sysdeps/unix/sysv/linux/raise\.c&#8203;:64
64    \.\./nptl/sysdeps/unix/sysv/linux/raise\.c&#8203;: No such file or directory\.

Test case hexdump&#8203;:
0000000 7271 2880 3f3f 0080                    
0000007

Debian 7\, kernel v3\.2\.63\-2\+deb7u2 x86\_64\, libc6 v2\.13\-38\+deb7u7\, GCC 4\.9\.2

**Comment by p5pRT** [comment]
### From @geeknik

[test95.pl](https://rt-archive.perl.org/perl5/Ticket/Attachment/1330053/710029/test95.pl)

**Comment by p5pRT** [comment]
### From @geeknik

This is a dupe of \#123755\. My apologies\, I thought I did a good enough job of searching before submission\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

This is a dupe of \#123755\. My apologies\, I thought I did a good enough job of searching before submission\.

**Comment by p5pRT** [comment]
### From @wolfsage

On Sun Feb 08 13&#8203;:11&#8203;:40 2015\, brian\.carpenter@&#8203;gmail\.com wrote&#8203;:
> This is a dupe of \#123755\. My apologies\, I thought I did a good enough
> job of searching before submission\.

No worries; I merged the two\, thanks\.

\-\- 
\-\- Matthew Horsfall \(alh\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @hvds

On Sat Feb 07 08&#8203;:27&#8203;:29 2015\, hv wrote&#8203;:
> % \./miniperl \-ce '/\(??/'
> miniperl&#8203;: sv\.c&#8203;:11434&#8203;: Perl\_sv\_vcatpvfn\_flags&#8203;: Assertion \`\(IV\)elen >= 0' failed\.
> Aborted \(core dumped\)
> % 

Code inspection shows similar problems for /\(?P/ and /\(?i\-/ \- in fact\, each time we use seqstart\.

> I think the diff below is probably the right fix\, I'll apply it along with
> a test in a couple of days if nobody disagrees\.
\[\.\.\.\]
> \-                   RExC\_parse\+\+;
> \+                    if \(\*RExC\_parse\)
> \+                        RExC\_parse\+\+;

This also needs to take UTF into account when skipping past an unknown char; is something like the below a reasonable way to test for that?
&nbsp;  \{
&nbsp;    eval qq\{/\(??\\x\{123\}/\};
&nbsp;    my $e = $@&#8203;;
&nbsp;    like\($e\, qr\{\\x\{123\}\}\, "Full utf8 character gets into error message"\);
&nbsp;  \}

Hugo

**Comment by p5pRT** [comment]
### From @hvds

Now fixed with 8a6d8ec6fe&#8203;:

&nbsp;    \[perl \#123755\] including unknown char in error requires care
&nbsp;    
&nbsp;    AFL \(\<http&#8203;://lcamtuf\.coredump\.cx/afl>\) found that when producing the
&nbsp;    error message for /\(??/ we hit an assert because we've stepped past
&nbsp;    the end of the pattern string\. Code inspection found that we also do
&nbsp;    that in other branches\, and we also need to check UTF more carefully\.

Hugo

**Comment by p5pRT** [comment]
@hvds - Status changed from 'open' to 'pending release'

**Comment by p5pRT** [comment]
### From @hvds

The new tests exposed an additional problem\, fixed in 7082c44063&#8203;:

&nbsp;    regcomp can read past end of string after parsing flags
&nbsp;    
&nbsp;    New test in 8a6d8ec6fe revealed additional code problem reading past
&nbsp;    end of string under clang with sanitize=address\.

Hugo

**Comment by p5pRT** [comment]
### From @khwilliamson

Thanks for submitting this ticket

The issue should be resolved with the release today of Perl v5\.22\, available at http&#8203;://www\.perl\.org/get\.html
If you find that the problem persists\, feel free to reopen this ticket

\-\- 
Karl Williamson for the Perl 5 porters team

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'pending release' to 'resolved'
